### PR TITLE
Record in/out transactions

### PR DESCRIPTION
**Pull request by IvanIvanoff**
Deployable but the influxdb db should be cleared once the etherbi transactions bug is fixed.

Fetch the transactions from etherbi and store them in time series data base.

Note: Has to be merged and deployed after https://github.com/santiment/sanbase2/pull/209

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -58,6 +58,12 @@ config :sanbase, Sanbase.ExternalServices.TwitterData.Store,
   pool: [ max_overflow: 10, size: 20 ],
   database: "twitter_followers_data"
 
+config :sanbase, Sanbase.Etherbi.Store,
+  host: {:system, "IFNLUXDB_HOST", "localhost"},
```
Comment (valo) [inline config/config.exs]: There is a typo in the name of the variable

File: lib/sanbase/etherbi/funds_movements.ex [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  @etherbi_url Config.module_get(Sanbase.Etherbi, :url)
+  @http_client Mockery.of("HTTPoison")
+
+  def transactions_in(wallets, from, to) do
+    from_unix = DateTime.to_unix(from, :seconds)
+    to_unix = DateTime.to_unix(to, :seconds)
+
+    url =
+      "#{@etherbi_url}/transactions_in?&from_timestamp=#{from_unix}&to_timestamp=#{to_unix}&wallets=#{
+        inspect(wallets)
```
Comment (valo) [inline lib/sanbase/etherbi/funds_movements.ex]: I think `Poison.encode!` will make more sense than using `inspect`. Also, passing JSON as GET parameter is a bit weird. Can't we have some escaping issues with this? Another problem with using GET parameters for encoding potentially long data is the URL length limit. I would suggest to use POST for such kind of data, which could potentially become quite long

File: lib/sanbase/etherbi/funds_movements.ex [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def transactions_out(wallets, from, to) do
+    from_unix = DateTime.to_unix(from, :seconds)
+    to_unix = DateTime.to_unix(to, :seconds)
+
+    wallets = Poison.encode!(wallets)
+
+    url =
+      "#{@etherbi_url}/transactions_in?from_timestamp=#{from_unix}&to_timestamp=#{to_unix}&wallets=#{
```
Comment (valo) [inline lib/sanbase/etherbi/funds_movements.ex]: It is possible to pass the GET parameters as `params` to the `get` function: https://hexdocs.pm/httpoison/HTTPoison.html#request/5

File: lib/sanbase/github/store.ex [hunk 4]
```diff
@@ -88,11 +88,14 @@ defmodule Sanbase.Github.Store do
...
-          |> Kernel.trunc()
-      end
+      activity =
+        if is_float(activity) do
+          activity
+            |> Float.ceil()
+            |> Kernel.trunc()
+        else
+          activity
+        end
```
Comment (valo) [inline lib/sanbase/github/store.ex]: Why the data should be truncated here?

Comment (IvanIvanoff) [reply lib/sanbase/github/store.ex]: This should not be in this PR

File: lib/sanbase/etherbi/funds_movements.ex (hunk 2 above)
Comment (IvanIvanoff) [reply lib/sanbase/etherbi/funds_movements.ex]: For now I'm only using it with one wallet, so it won't get big. For multiple wallets passing a single from and to timestamps won't work well and has to be reworked

File: lib/sanbase/etherbi/store.ex [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    |> parse_transactions_time_series()
+  end
+
+  defp transactions_from_to_query(measurement, from, to, _resolution, transaction_type) do
+
+    ~s/SELECT volume, token FROM "#{measurement}"
+    WHERE transaction_type = '#{transaction_type}'
+    AND time >= #{DateTime.to_unix(from, :nanoseconds)}
+    AND time <= #{DateTime.to_unix(to, :nanoseconds)}/
+    # GROUP BY time(#{resolution}) fill(none)/
```
Comment (valo) [inline lib/sanbase/etherbi/store.ex]: Let's not keep commented out code around

File: lib/sanbase/etherbi/transactions.ex [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+  def init(:ok) do
+    if Config.get(:sync_enabled, false) do
+      Store.create_db()
+
+      update_interval_ms = Config.get(:update_interval, @default_update_interval)
+      token_decimals = build_token_decimals_map()
+
+      GenServer.cast(self(), :sync)
+      {:ok, %{update_interval_ms: update_interval_ms, token_decimals: token_decimals}}
```
Comment (valo) [inline lib/sanbase/etherbi/transactions.ex]: If we keep the `token_decimals` in the state of the process, what is going to happen if a new project gets added through the admin panel?

Comment (IvanIvanoff) [reply lib/sanbase/etherbi/transactions.ex]: We can easily rebuild this map in `handle_cast`. The bigger problem I see though is when we add a new project what will we do with the past transactions. We would check for transactions after `last_datetime`, but there would be transactions before that time with that new token.

File: lib/sanbase/etherbi/transactions.ex [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  # number of decimal places `nil` is written instead and it gets filtered by the Store.import()
+  defp convert_to_measurement(
+         transactions_data,
+         measurement_name,
+         transaction_type,
+         token_decimals
+       ) do
+    transactions_data
+    |> Enum.map(fn {datetime, volume, _address, token} ->
+      if decimal_places = Map.has_key?(token_decimals, token) do
```
Comment (valo) [inline lib/sanbase/etherbi/transactions.ex]: Can we filter out the tokens without decimals before fetching the transactions and avoiding fetching the transactions at all? That should remove this `if`

File: lib/sanbase_web/graphql/schema.ex [hunk 8]
```diff
@@ -129,6 +129,16 @@ defmodule SanbaseWeb.Graphql.Schema do
       resolve(&EtherbiResolver.transaction_volume/3)
     end
 
+    @desc "Shows the flow of funds in an exchange wallet"
+    field :transactions, list_of(:transaction) do
+      arg(:wallet, non_null(:string))
+      arg(:from, non_null(:datetime))
+      arg(:to, :datetime, default_value: DateTime.utc_now())
+      arg(:interval, :string, default_value: "1h")
+      arg(:transaction_type, :string)
```
Comment (valo) [inline lib/sanbase_web/graphql/schema.ex]: Can we use an `enum` for this? https://hexdocs.pm/absinthe/Absinthe.Schema.Notation.html#enum/3

File: lib/sanbase/etherbi/store.ex (hunk 5 above)
Comment (IvanIvanoff) [reply lib/sanbase/etherbi/store.ex]: done

File: lib/sanbase_web/graphql/schema.ex (hunk 8 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/schema.ex]: I tried but I did not like that it has to be used without quotes and uppercase. It just differs too much from everything else. Example: `transaction_type: IN`. Though, that's only a style preference and could be changed if found to be the better option

**Comment by valo** [comment]
What is remaining to be done on this one?

**Comment by IvanIvanoff** [comment]
Wrong transactions will be fetched from etherbi if we deploy before the bugfix is applied there

**Comment by valo** [comment]
Can we deploy this and then clear the influxDB cache after the transactions fix is in place?

**Comment by IvanIvanoff** [comment]
Yes, we can

File: lib/sanbase/etherbi/transactions.ex [hunk 9]
```diff
@@ -0,0 +1,131 @@
+defmodule Sanbase.Etherbi.Transactions do
+  @moduledoc ~S"""
+    This module is a GenServer that periodically sends requests to etherbi API.
+    In and out transactions are fetched and saved in a time series database for
+    easier aggregation and querying.
+  """
```
Comment (valo) [inline lib/sanbase/etherbi/transactions.ex]: Can we add a function here, which will purge the whole cache? We will need this function when we add new exchange wallets and when we fix bugs, like the invalid transaction filtering bug.

Comment (IvanIvanoff) [reply lib/sanbase/etherbi/transactions.ex]: There are `list_measurements` and `drop_measurement` in the `Store` module, but they are intended to be used in migrations. Do you mean such purging or one that can happen at runtime?

Comment (valo) [reply lib/sanbase/etherbi/transactions.ex]: The use case will be:

1. Garry adds/removes a new exchange wallet in the admin panel
2. The cache is cleared automatically and starts repopulating immediately

I am not sure how we will hook up the add/remove event from the admin to he cache purge, but in any case we will need some function to call to do it.

Comment (IvanIvanoff) [reply lib/sanbase/etherbi/transactions.ex]: There are `list_measurements` and `drop_measurement` in the `Store` module which maybe can be used in a migration. Do you mean this kind of purge or one that can happen on runtime? For the later we can develop into a complete tool the Elixir mini-project I used to import the data from staging to prod without downtime.

Comment (valo) [reply lib/sanbase/etherbi/transactions.ex]: That will work. So no need for a special function

**Pull request merged**